File: frontend/src/services/planService.js
import { getStore, setStore, generateId, delay, getCurrentUserId, buildPlanForType, KEYS } from './mockData'
import { isOpenAIAvailable, generateWithOpenAI } from './openaiService'

function flattenPlan(plan) {
  if (!plan) return plan
  const { generated_plan, ...rest } = plan
  return { ...rest, ...generated_plan }
}

function getUserProfile() {
  const userId = getCurrentUserId()
  if (!userId) return null
  const profiles = getStore(KEYS.PROFILES) || []
  return profiles.find((p) => p.user_id === userId) || null
}

export const generatePlan = async (activityInput) => {
  const userId = getCurrentUserId()
  if (!userId) throw { response: { status: 401, data: { error: 'Not authenticated' } } }

  const startTime = Date.now()
  let generatedPlan
  let modelUsed

  if (isOpenAIAvailable()) {
    try {
      const profile = getUserProfile()
      generatedPlan = await generateWithOpenAI(activityInput, profile)
      modelUsed = 'gpt-4o-mini'
    } catch (err) {
      console.warn('OpenAI call failed, falling back to mock templates:', err.message)
      generatedPlan = buildPlanForType(activityInput.activity_type || 'custom', activityInput)
      modelUsed = 'mock-local (fallback)'
    }
  } else {
    await delay(800)
    generatedPlan = buildPlanForType(activityInput.activity_type || 'custom', activityInput)
    modelUsed = 'mock-local'
  }

  const plan = {
    _id: generateId(),
    user_id: userId,
    activity_input: activityInput,
    generated_plan: generatedPlan,
    ai_model_used: modelUsed,
    generation_time_ms: Date.now() - startTime,
    created_at: new Date().toISOString(),
  }

  const plans = getStore(KEYS.PLANS) || []
  plans.unshift(plan)
  setStore(KEYS.PLANS, plans)

  return { data: { plan: flattenPlan(plan), message: 'Plan generated successfully' } }
}

export const getPlan = async (planId) => {
  await delay(200)
  const plans = getStore(KEYS.PLANS) || []
  const plan = plans.find((p) => p._id === planId)
  if (!plan) throw { response: { status: 404, data: { error: 'Plan not found' } } }

  return { data: { plan: flattenPlan(plan) } }
}

export const getPlanHistory = async (page = 1, perPage = 10) => {
  await delay(200)
  const userId = getCurrentUserId()
  if (!userId) throw { response: { status: 401, data: { error: 'Not authenticated' } } }

  const allPlans = (getStore(KEYS.PLANS) || []).filter((p) => p.user_id === userId)
  allPlans.sort((a, b) => new Date(b.created_at) - new Date(a.created_at))

  const total = allPlans.length
  const start = (page - 1) * perPage
  const paged = allPlans.slice(start, start + perPage)

  const plans = paged.map((p) => ({
    _id: p._id,
    title: p.generated_plan?.title || 'Untitled Plan',
    summary: p.generated_plan?.summary || '',
    activity_type: p.activity_input?.activity_type || p.generated_plan?.activity_type || 'custom',
    created_at: p.created_at,
    date: p.activity_input?.date,
  }))

  return {
    data: {
      plans,
      total,
      page,
      per_page: perPage,
      total_pages: Math.ceil(total / perPage) || 1,
    },
  }
}

export const deletePlan = async (planId) => {
  await delay(200)
  const plans = getStore(KEYS.PLANS) || []
  const idx = plans.findIndex((p) => p._id === planId)
  if (idx === -1) throw { response: { status: 404, data: { error: 'Plan not found' } } }

  plans.splice(idx, 1)
  setStore(KEYS.PLANS, plans)

  return { data: { message: 'Plan deleted successfully' } }
}

File: backend/app/services/plan_service.py
import time

from app.models.plan import PlanModel
from app.models.profile import ProfileModel
from app.ai.engine import AIEngine
from app.utils.helpers import serialize_doc
from app.utils.validators import validate_activity_type


class PlanService:

    @staticmethod
    def generate_plan(user_id, activity_input):
        activity_type = activity_input.get('activity_type', 'custom')
        if not validate_activity_type(activity_type):
            return {'error': f'Invalid activity type: {activity_type}'}, 400

        profile = ProfileModel.get_by_user_id(user_id)
        if not profile:
            return {'error': 'Please create a profile before generating plans'}, 400

        try:
            engine = AIEngine()
            result = engine.generate_plan(profile, activity_input)

            plan_id = PlanModel.create_plan(
                user_id=user_id,
                activity_input=activity_input,
                generated_plan=result['plan'],
                ai_model_used=result['model_used'],
                generation_time_ms=result['generation_time_ms'],
            )

            plan = PlanModel.get_by_id(plan_id)
            return {
                'message': 'Plan generated successfully',
                'plan': serialize_doc(plan),
            }, 201

        except Exception as e:
            return {'error': f'Failed to generate plan: {str(e)}'}, 500

    @staticmethod
    def get_plan(plan_id, user_id):
        plan = PlanModel.get_by_id(plan_id)
        if not plan:
            return {'error': 'Plan not found'}, 404
        if str(plan['user_id']) != user_id:
            return {'error': 'Unauthorized'}, 403
        return {'plan': serialize_doc(plan)}, 200

    @staticmethod
    def get_user_plans(user_id, page=1, per_page=10):
        result = PlanModel.get_user_plans(user_id, page, per_page)
        result['plans'] = [serialize_doc(p) for p in result['plans']]
        return result, 200

    @staticmethod
    def delete_plan(plan_id, user_id):
        deleted = PlanModel.delete_plan(plan_id, user_id)
        if not deleted:
            return {'error': 'Plan not found or unauthorized'}, 404
        return {'message': 'Plan deleted successfully'}, 200
